tsl/src/continuous_aggs/materialize.c

```diff
@@ -43,33 +46,113 @@ static char *build_merge_update_clause(List *column_names);
-										  SchemaAndName partial_view,
-										  SchemaAndName materialization_table,
-										  const NameData *time_column_name,
-										  TimeRange materialization_range,
-										  const char *const chunk_condition);
-static bool spi_exists_materializations(SchemaAndName materialization_table,
-										const NameData *time_column_name,
-										TimeRange materialization_range);
-static uint64 spi_merge_materializations(Hypertable *mat_ht, const ContinuousAgg *cagg,
-										 SchemaAndName partial_view,
-										 SchemaAndName materialization_table,
-										 const NameData *time_column_name,
-										 TimeRange materialization_range);
-
+typedef enum MaterializationPlanType
+{
+	PLAN_TYPE_INSERT,
+	PLAN_TYPE_DELETE,
+	PLAN_TYPE_EXISTS,
+	PLAN_TYPE_MERGE,
+	PLAN_TYPE_MERGE_DELETE,
+	_MAX_MATERIALIZATION_PLAN_TYPES
+} MaterializationPlanType;
+
+typedef struct MaterializationContext
+{
+	Hypertable *mat_ht;
+	const ContinuousAgg *cagg;
+	SchemaAndName partial_view;
+	SchemaAndName materialization_table;
+	NameData *time_column_name;
+	TimeRange materialization_range;
+	char *chunk_condition;
+} MaterializationContext;
+
+typedef char *(*MaterializationCreateStatement)(MaterializationContext *context);
+typedef void (*MaterializationEmitError)(MaterializationContext *context);
+typedef void (*MaterializationEmitProgress)(MaterializationContext *context, uint64 rows_processed);
+
+typedef struct MaterializationPlan
+{
+	const char *query;
+	SPIPlanPtr plan;
+	bool read_only;
+	MaterializationCreateStatement create_statement;
+	MaterializationEmitError emit_error;
+	MaterializationEmitProgress emit_progress;
+} MaterializationPlan;
+
+static char *create_materialization_insert_statement(MaterializationContext *context);
+static char *create_materialization_delete_statement(MaterializationContext *context);
+static char *create_materialization_exists_statement(MaterializationContext *context);
+static char *create_materialization_merge_statement(MaterializationContext *context);
+static char *create_materialization_merge_delete_statement(MaterializationContext *context);
+
+static void emit_materialization_insert_error(MaterializationContext *context);
+static void emit_materialization_delete_error(MaterializationContext *context);
+static void emit_materialization_merge_error(MaterializationContext *context);
+
+static void emit_materialization_insert_progress(MaterializationContext *context,
+												 uint64 rows_processed);
+static void emit_materialization_delete_progress(MaterializationContext *context,
+												 uint64 rows_processed);
+static void emit_materialization_merge_progress(MaterializationContext *context,
+												uint64 rows_processed);
+
+static MaterializationPlan materialization_plans[_MAX_MATERIALIZATION_PLAN_TYPES + 1] = {
+	[PLAN_TYPE_INSERT] = { .query = "INSERT INTO %s.%s SELECT * FROM %s.%s AS I "
+									"WHERE I.%s >= $1 AND I.%s < $2 %s;",
```

Comment:
Nit: Even though you can store the query string side-by-side with the function pointers, the string is static and you can just as easily add the string literal inside the function.

Reply:
Make sense, added the string literal inside the correspondent functions.